File: Server/Server/script/db/DbManager.cs
using System;
using System.Text.RegularExpressions;
using System.Collections.Generic;
using MongoDB.Driver;
using MongoDB.Bson;
using MongoDB.Bson.Serialization;
using DBEntity;

public class DBManager
{


    public static IMongoDatabase database;


    /// <summary>
    /// 连接mongoDB
    /// </summary>
    /// <param name="dbname">数据库名称</param>
    /// <param name="connStr">连接URL</param>
    /// <returns>是否连接成功</returns>
    public static bool Connect(string dbname, string connStr) {
        //连接
        try {
            //获得连接
            var client = new MongoClient(connStr);
            //获得数据库
            database = client.GetDatabase(dbname);
            Console.WriteLine("[数据库]connect 成功 ");
            return true;
        }
        catch (Exception e) {
            Console.WriteLine("[数据库]connect 失败, " + e.Message);
            return false;
        }
    }



    /// <summary>
    /// 测试并重连
    /// </summary>
    private static void CheckAndReconnect() {
        try {
            if (database != null) {
                return;
            }
            Connect("game", "mongodb://127.0.0.1:27017");
            Console.WriteLine("[数据库] Reconnect!");
        }
        catch (Exception e) {
            Console.WriteLine("[数据库] CheckAndReconnect 失败 " + e.Message);
        }

    }

    /// <summary>
    /// 判定安全字符串
    /// </summary>
    /// <param name="str"></param>
    /// <returns></returns>
    private static bool IsSafeString(string str) {
        return !Regex.IsMatch(str, @"[-|;|,|\/|\(|\)|\[|\]|\}|\{|%|@|\*|!|\']");
    }


    /// <summary>
    /// 是否存在该用户 若为真，不存在该账户
    /// </summary>
    /// <param name="id"></param>
    /// <returns></returns>
    public static bool IsAccountExist(string uid) {
        Console.WriteLine(uid);
        CheckAndReconnect();
        //防注入
        if (!IsSafeString(uid)) {
            return false;
        }
        //创建约束生成器
        FilterDefinitionBuilder<BsonDocument> builder = Builders<BsonDocument>.Filter;
        //约束条件
        FilterDefinition<BsonDocument> filter = builder.Eq("uid", uid);
        //查询
        try {
            //获取数据
            IMongoCollection<BsonDocument> coll = database.GetCollection<BsonDocument>("account");
            var result = coll.Find(filter).ToList();
            return result.Count == 0 ? false : true;
        }
        catch (Exception e) {
            Console.WriteLine("[数据库] IsSafeString err, " + e.Message);
            return false;
        }
    }

    //注册
    public static bool Register(string uid, string pw) {
        CheckAndReconnect();
        //防sql注入
        if (!IsSafeString(uid)) {
            Console.WriteLine("[数据库] Register fail, id not safe");
            return false;
        }
        if (!IsSafeString(pw)) {
            Console.WriteLine("[数据库] Register fail, pw not safe");
            return false;
        }
        //能否注册
        if (IsAccountExist(uid)) {
            Console.WriteLine("[数据库] Register fail, id exist");
            return false;
        }
        //写入数据库User表
        Account account = new Account();
        account.uid = uid;
        account.pw = pw;
        try {
            IMongoCollection<MongoBaseEntity> coll = database.GetCollection<MongoBaseEntity>("account");
            coll.InsertOne(account);
            return true;
        }
        catch (Exception e) {
            Console.WriteLine("[数据库] Register fail " + e.Message);
            return false;
        }
    }

    //创建角色
    public static bool CreatePlayer(string uid) {
        CheckAndReconnect();
        //防注入
        if (!IsSafeString(uid)) {
            Console.WriteLine("[数据库] CreatePlayer fail, id not safe");
            return false;
        }
        DBEntity.Player player = new DBEntity.Player();
        player.uid = uid;
        player.headPhoto = 0;
        player.palyerIntroduction = "这个人很懒，什么也没有写。";
        try {
            IMongoCollection<MongoBaseEntity> coll = database.GetCollection<MongoBaseEntity>("player");
            coll.InsertOne(player);
            return true;
        }
        catch (Exception e) {
            Console.WriteLine("[数据库] CreatePlayer err, " + e.Message);
            return false;
        }
    }


    //检测用户名密码
    public static bool CheckPassword(string uid, string pw) {
        CheckAndReconnect();
        //防注入
        if (!IsSafeString(uid)) {
            Console.WriteLine("[数据库] CheckPassword fail, id not safe");
            return false;
        }
        if (!IsSafeString(pw)) {
            Console.WriteLine("[数据库] CheckPassword fail, pw not safe");
            return false;
        }
        //查询
        //创建约束生成器
        FilterDefinitionBuilder<BsonDocument> builder = Builders<BsonDocument>.Filter;
        //约束条件
        FilterDefinition<BsonDocument> filter = builder.And(builder.Eq("uid", uid), builder.Eq("pw", pw));
        
        try {
            //获取数据
            IMongoCollection<BsonDocument> coll = database.GetCollection<BsonDocument>("account");
            var result = coll.Find(filter).ToList();
            return result.Count == 0 ? false : true;
        }
        catch (Exception e) {
            Console.WriteLine("[数据库] CheckPassword err, " + e.Message);
            return false;
        }
    }


    //获取玩家数据
    public static PlayerData GetPlayerData(string uid) {
        CheckAndReconnect();
        //防注入
        if (!IsSafeString(uid)) {
            Console.WriteLine("[数据库] GetPlayerData fail, id not safe");
            return null;
        }
        //查询
        //创建约束生成器
        FilterDefinitionBuilder<BsonDocument> builder = Builders<BsonDocument>.Filter;
        //约束条件
        FilterDefinition<BsonDocument> filter = builder.Eq("uid", uid);
        try {
            //查询
            //获取数据
            IMongoCollection<BsonDocument> coll = database.GetCollection<BsonDocument>("player");
            var result = coll.Find(filter).ToList();
            if (!(result.Count == 0 ? false : true)) {
                return null;
            }
            //反序列化
            DBEntity.Player player = BsonSerializer.Deserialize<DBEntity.Player>(result[0]);

            PlayerData playerData = new PlayerData();
            playerData.headPhoto = player.headPhoto;
            playerData.palyerIntroduction = player.palyerIntroduction;
            return playerData;
        }
        catch (Exception e) {
            Console.WriteLine("[数据库] GetPlayerData fail, " + e.Message);
            return null;
        }
    }


    //保存角色
    public static bool UpdatePlayerData(string uid, PlayerData playerData) {
        CheckAndReconnect();
        //创建约束生成器
        FilterDefinitionBuilder<BsonDocument> builder1 = Builders<BsonDocument>.Filter;
        //约束条件
        FilterDefinition<BsonDocument> filter = builder1.Eq("uid", uid);
        UpdateDefinitionBuilder<BsonDocument> builder2 = Builders<BsonDocument>.Update;
        UpdateDefinition<BsonDocument> update = builder2.Combine(builder2.Set("headPhoto", playerData.headPhoto), builder2.Set("palyerIntroduction", playerData.palyerIntroduction));
        //更新
        try {
            //获取数据
            IMongoCollection<BsonDocument> coll = database.GetCollection<BsonDocument>("player");
            coll.UpdateOne(filter,update);
            return true;
        }
        catch (Exception e) {
            Console.WriteLine("[数据库] UpdatePlayerData err, " + e.Message);
            return false;
        }
    }

    //获取好友列表
    public static List<string> GetFriendList(string uid) {
        CheckAndReconnect();
        //防sql注入
        if (!IsSafeString(uid)) {
            Console.WriteLine("[数据库] GetPlayerData fail, id not safe");
            return null;
        }
        //创建约束生成器
        FilterDefinitionBuilder<BsonDocument> builder = Builders<BsonDocument>.Filter;
        //约束条件
        FilterDefinition<BsonDocument> filter = builder.Eq("uid", uid);
        try {
            //查询
            //获取数据
            IMongoCollection<BsonDocument> coll = database.GetCollection<BsonDocument>("friend");
            var result = coll.Find(filter).ToList();
            if (!(result.Count == 0 ? false : true)) {
                return null;
            }
            //读取
            List<string> friendList = new List<string>();
            foreach (var item in result) {
                friendList.Add(BsonSerializer.Deserialize<Friend>(item).friendId);
            }
            return friendList;
        }
        catch (Exception e) {
            Console.WriteLine("[数据库] GetData fail, " + e.Message);
            return null;
        }
    }

    //删除好友
    public static bool DeleteFriend(string uid, string friendId) {
        CheckAndReconnect();
        //防注入
        if (!IsSafeString(uid)) {
            Console.WriteLine("[数据库] DeleteFriend fail, id not safe");
            return false;
        }
        if (!IsSafeString(friendId)) {
            Console.WriteLine("[数据库] DeleteFriend fail, friendId not safe");
            return false;
        }
        //创建约束生成器
        FilterDefinitionBuilder<BsonDocument> builder = Builders<BsonDocument>.Filter;
        //约束条件
        FilterDefinition<BsonDocument> filter1 = builder.And(builder.Eq("uid", uid), builder.Eq("friendid", friendId));
        FilterDefinition<BsonDocument> filter2 = builder.And(builder.Eq("uid", friendId), builder.Eq("friendid", uid));
        try {
            //获取数据
            IMongoCollection<BsonDocument> coll = database.GetCollection<BsonDocument>("friend");
            coll.DeleteMany(filter1);
            coll.DeleteMany(filter2);
            return true;
        }
        catch (Exception e) {
            Console.WriteLine("[数据库] DeleteFriend err, " + e.Message);
            return false;
        }
    }

    //是否是好友
    public static bool IsFriendExist(string uid, string friendId) {
        CheckAndReconnect();
        //防注入
        if (!IsSafeString(uid)) {
            return false;
        }
        if (!IsSafeString(friendId)) {
            return false;
        }
        //创建约束生成器
        FilterDefinitionBuilder<BsonDocument> builder = Builders<BsonDocument>.Filter;
        //约束条件
        FilterDefinition<BsonDocument> filter = builder.And(builder.Eq("uid", uid), builder.Eq("friendid", friendId));
        //查询
        try {
            IMongoCollection<BsonDocument> coll = database.GetCollection<BsonDocument>("friend");
            var result = coll.Find(filter).ToList();
            return result.Count == 0 ? false : true;
        }
        catch (Exception e) {
            Console.WriteLine("[数据库] IsSafeString err, " + e.Message);
            return false;
        }
    }

    //添加好友
    public static bool AddFriend(string uid, string friendId) {
        CheckAndReconnect();
        //防sql注入
        if (!IsSafeString(uid)) {
            Console.WriteLine("[数据库] AddFriend fail, id not safe");
            return false;
        }
        if (!IsSafeString(friendId)) {
            Console.WriteLine("[数据库] AddFriend fail, friendId not safe");
            return false;
        }

        Friend friend1 = new Friend();
        friend1.uid = uid;
        friend1.friendId = friendId;
        Friend friend2 = new Friend();
        friend2.uid = friendId;
        friend2.friendId = uid;

        try {
            IMongoCollection<MongoBaseEntity> coll = database.GetCollection<MongoBaseEntity>("friend");
            coll.InsertOne(friend1);
            coll.InsertOne(friend2);
            return true;
        }
        catch (Exception e) {
            Console.WriteLine("[数据库] AddFriend err, " + e.Message);
            return false;
        }
    }
}

File: Server-Go/Demo/db/DBManager.go
package db

import (
	"ReaminHistory/Demo/Helper"
	"ReaminHistory/Demo/Player/PlayerData"
	"context"
	"fmt"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
	"go.mongodb.org/mongo-driver/mongo/readpref"
	"strings"
	"time"
)

var database *mongo.Database

//连接数据库
func Connect(dbname string, connStr string) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()
	client, err := mongo.Connect(ctx, options.Client().ApplyURI(connStr))
	if err != nil {
		panic(err)
	}
	// Ping the primary
	if err := client.Ping(ctx, readpref.Primary()); err != nil {
		fmt.Println("连接数据库成功失败")
		panic(err)
	}
	database = client.Database(dbname)
	fmt.Println("连接数据库成功", database.Name())
}

//测试并重连
func CheckAndReconnect() {
	if database == nil {
		netInfo := Helper.GetNetConfig()
		Connect(netInfo.DBName, netInfo.DBURL)
		fmt.Println("数据库重连")
	}
}

//判断安全字符
func IsSafeString(str string) bool {
	s := [...]string{"[", "-", "|", ",", ".", "!", "@", "#", "$", "%", "^", "&", "*", "(", ")", "_", "=", "+", "]", "{", "}", "'", ":", ";", "?", "/"}
	for _, v := range s {
		if strings.ContainsAny(str, v) {
			return false
		}
	}
	return true
}

//是否存在该用户
func IsAccountExist(id string) bool {
	CheckAndReconnect()
	if !IsSafeString(id) {
		return false
	}
	account := new(Account)
	account.UId = id
	query := database.Collection("account").FindOne(context.TODO(), account)
	res, err := query.DecodeBytes()
	if err == nil && res.String() != "" {
		return true
	} else {
		return false
	}
}

//注册
func Register(uid string, pw string) bool {
	CheckAndReconnect()
	//防止特殊字符
	if !IsSafeString(uid) {
		fmt.Println("[数据库] Register fail, id not safe")
		return false
	}
	if !IsSafeString(pw) {
		fmt.Println("[数据库] Register fail, pw not safe")
		return false
	}
	//uid是否存在
	if IsAccountExist(uid) {
		fmt.Println("[数据库] Register fail, id exist")
		return false
	}
	//写入数据库
	account := new(Account)
	account.UId = uid
	account.Pw = pw
	_, err := database.Collection("account").InsertOne(context.TODO(), account)
	if err != nil {
		fmt.Println("[数据库] Register fail")
		return false
	} else {
		return true
	}
}

//创建角色
func CreatePlayer(uid string) bool {
	CheckAndReconnect()
	//防止特殊字符
	if !IsSafeString(uid) {
		fmt.Println("[数据库] CreatePlayer fail, id not safe")
		return false
	}
	player := new(Player)
	player.UId = uid
	player.HeadPhoto = 0
	player.PlayerIntroduction = "这个人很懒，什么也没有写。"
	_, err := database.Collection("account").InsertOne(context.TODO(), player)
	if err != nil {
		fmt.Println("[数据库] CreatePlayer fail")
		return false
	} else {
		return true
	}
}

//检测用户名及密码
func CheckPassword(uid string, pw string) bool {
	CheckAndReconnect()
	//防止特殊字符
	if !IsSafeString(uid) {
		fmt.Println("[数据库] CheckPassword fail, id not safe")
		return false
	}
	if !IsSafeString(pw) {
		fmt.Println("[数据库] CheckPassword fail, pw not safe")
		return false
	}
	account := new(Account)
	account.UId = uid
	account.Pw = pw
	query := database.Collection("account").FindOne(context.TODO(), account)
	res, err := query.DecodeBytes()
	if err == nil && res.String() != "" {
		return true
	} else {
		return false
	}
}

//获取玩家数据
func GetPlayerData(uid string) *PlayerData.PlayerData {
	CheckAndReconnect()
	//防止特殊字符
	if !IsSafeString(uid) {
		fmt.Println("[数据库] GetPlayerData fail, id not safe")
		return nil
	}
	player := new(Player)
	player.UId = uid
	query := database.Collection("player").FindOne(context.TODO(), player)
	err := query.Decode(player)
	if err != nil {
		fmt.Println("[数据库] GetPlayerData fail")
		return nil
	}
	playerData := new(PlayerData.PlayerData)
	playerData.HeadPhoto = player.HeadPhoto
	playerData.PlayerIntroduction = player.PlayerIntroduction
	return playerData
}

//保存玩家角色数据
func UpdatePlayerData(uid string, data *PlayerData.PlayerData) bool {
	CheckAndReconnect()
	//防止特殊字符
	if !IsSafeString(uid) {
		fmt.Println("[数据库] UpdatePlayerData fail, id not safe")
		return false
	}
	player := new(Player)
	player.UId = uid
	player.HeadPhoto = data.HeadPhoto
	player.PlayerIntroduction = data.PlayerIntroduction
	update := bson.D{{"$set", player}}
	res := database.Collection("player").FindOneAndUpdate(context.TODO(), bson.M{"uid": uid}, update)
	query, err := res.DecodeBytes()
	if err != nil || query.String() == "" {
		return false
	}
	return true
}

//获取好友列表
func GetFriendList(uid string) string {
	CheckAndReconnect()
	//防止特殊字符
	if !IsSafeString(uid) {
		fmt.Println("[数据库] GetFriendList fail, id not safe")
		return ""
	}
	query, err := database.Collection("friend").Find(context.TODO(), bson.M{"uid": uid})
	if err != nil {
		fmt.Println("[数据库] GetFriendList fail")
		return ""
	}
	friends := make([]Friend, query.RemainingBatchLength())
	err1 := query.All(context.TODO(), &friends)
	if err1 != nil {
		fmt.Println("[数据库] GetFriendList fail")
		return ""
	}
	friendList := ""
	for _, s := range friends {
		friendList += s.FriendId + ","
	}
	return friendList
}

//删除好友
func DeleteFriend(uid string, friendId string) bool {
	CheckAndReconnect()
	//防止特殊字符
	if !IsSafeString(uid) {
		fmt.Println("[数据库] DeleteFriend fail, id not safe")
		return false
	}
	if !IsSafeString(friendId) {
		fmt.Println("[数据库] DeleteFriend fail, friendId not safe")
		return false
	}
	friend1 := new(Friend)
	friend1.UId = uid
	friend1.FriendId = friendId
	friend2 := new(Friend)
	friend2.FriendId = uid
	friend2.UId = friendId
	_, err1 := database.Collection("friend").DeleteMany(context.TODO(), friend1)
	_, err2 := database.Collection("friend").DeleteMany(context.TODO(), friend2)
	if err1 != nil || err2 != nil {
		return false
	} else {
		return true
	}
}

//是否是好友
func IsFriendExist(uid string, friendId string) bool {
	CheckAndReconnect()
	//防止特殊字符
	if !IsSafeString(uid) {
		fmt.Println("[数据库] IsFriendExist fail, id not safe")
		return false
	}
	if !IsSafeString(friendId) {
		fmt.Println("[数据库] IsFriendExist fail, id not safe")
		return false
	}
	friend := new(Friend)
	friend.UId = uid
	friend.FriendId = friendId
	query := database.Collection("friend").FindOne(context.TODO(), friend)
	res, err := query.DecodeBytes()
	if err == nil && res.String() != "" {
		return true
	} else {
		return false
	}
}

//添加好友
func AddFriend(uid string, friendId string) bool {
	CheckAndReconnect()
	//防止特殊字符
	if !IsSafeString(uid) {
		fmt.Println("[数据库] AddFriend fail, id not safe")
		return false
	}
	if !IsSafeString(friendId) {
		fmt.Println("[数据库] AddFriend fail, id not safe")
		return false
	}
	if IsFriendExist(uid, friendId) {
		fmt.Println("[数据库] AddFriend fail, id exit")
		return false
	}
	if IsFriendExist(friendId, uid) {
		fmt.Println("[数据库] AddFriend fail, id exit")
		return false
	}
	friend1 := new(Friend)
	friend1.UId = uid
	friend1.FriendId = friendId
	friend2 := new(Friend)
	friend2.FriendId = uid
	friend2.UId = friendId
	//写入数据库
	_, err1 := database.Collection("friend").InsertOne(context.TODO(), friend1)
	_, err2 := database.Collection("friend").InsertOne(context.TODO(), friend2)
	if err1 != nil || err2 != nil {
		fmt.Println("[数据库] AddFriend fail")
		return false
	} else {
		return true
	}
}
